lib/crypto/packaged.py

```diff
@@ -104,6 +113,34 @@ def sign_file(file_obj):
     return file_obj
 
 
+def sign_multi(file_path, endpoint, guid):
+    """Sign the internal extensions from a multi-package XPI (if any)."""
+    log.info('Signing multi-package file {0}'.format(file_path))
+    # Extract the multi-package to a temp folder.
+    folder = tempfile.mkdtemp()
+    try:
+        extract_xpi(file_path, folder)
+        xpis = glob.glob(os.path.join(folder, '*.xpi'))
+        cert_serial_nums = []  # The certificate serial numbers for the XPIs.
+        for xpi in xpis:
+            xpi_info = parse_xpi(xpi, check=False)
+            if xpi_info['type'] == amo.ADDON_EXTENSION:
+                # Sign the internal extensions in place.
+                cert_serial_nums.append(call_signing(xpi, endpoint, guid))
+        # Repackage (re-zip) the multi-package.
+        # We only want the (unique) temporary file name.
+        with tempfile.NamedTemporaryFile() as temp_file:
+            temp_filename = temp_file.name
+        shutil.make_archive(temp_filename, format='zip', root_dir=folder)
+        # Now overwrite the current multi-package xpi with the repackaged one.
+        # Note that shutil.make_archive automatically added a '.zip' to the end
+        # of the base_name provided as first argument.
+        shutil.move('{0}.zip'.format(temp_filename), file_path)
+        return '\n'.join([str(num) for num in cert_serial_nums])
+    finally:
+        amo.utils.rm_local_tmp_dir(folder)
```

Comment:
It's getting more hairy by the day, and I don't like that, but didn't find a nicer/shorter/clearer way to do it.